components/normalizer/src/lib.rs

```diff
@@ -647,32 +643,28 @@ where
         &mut self,
         low: u16,
         offset: usize,
-        slice32: &ZeroSlice<U24>,
+        slice32: &ZeroSlice<char>,
     ) -> (char, usize) {
         let len = usize::from(low >> 13) + 1;
         let (starter, tail) = slice32
             .get_subslice(offset..offset + len)
             .and_then(|slice| slice.split_first())
-            .map_or_else(
-                || {
-                    // GIGO case
-                    debug_assert!(false);
-                    (REPLACEMENT_CHARACTER, EMPTY_U24)
-                },
-                |(first, trail)| (char_from_u24(first), trail),
-            );
+            .unwrap_or_else(|| {
+                // GIGO case
+                debug_assert!(false);
+                (REPLACEMENT_CHARACTER, EMPTY_CHAR)
+            });
         if low & 0x1000 != 0 {
             // All the rest are combining
-            for u in tail.iter() {
+            for ch in tail.iter() {
                 self.buffer
-                    .push(CharacterAndClass::new_with_placeholder(char_from_u24(u)));
+                    .push(CharacterAndClass::new_with_placeholder(ch));
             }
```

Comment:
Thought for later: using `Vec::extend` has better performance than repeated `Vec::push`, because it can preallocate and vectorize better. You could try rewriting your `push`ing for-loops with `extends`, maybe there's some performance to be gained:
```rust
self.buffer.extend(tail.iter().map(CharacterAndClass::new_with_placeholder));
```

Reply:
Thanks. [Filed.](https://github.com/unicode-org/icu4x/issues/2482)